libs/mock-builder/src/storage.rs

```diff
@@ -1,64 +1,87 @@
+
+use super::util::TypeSignature;
 
 /// Identify a call in the call storage
 pub type CallId = u64;
 
-trait Callable {
-	fn as_any(&self) -> &dyn Any;
+struct CallInfo {
+	ptr: u128,
+	type_signature: TypeSignature,
 }
 
 thread_local! {
-	static CALLS: RefCell<HashMap<CallId, Box<dyn Callable>>>
-		= RefCell::new(HashMap::default());
-}
-
-struct CallWrapper<Input, Output>(Box<dyn Fn(Input) -> Output>);
-
-impl<Input: 'static, Output: 'static> Callable for CallWrapper<Input, Output> {
-	fn as_any(&self) -> &dyn Any {
-		self
-	}
+	static CALLS: RefCell<HashMap<CallId, CallInfo>> = RefCell::new(HashMap::default());
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Error {
 	CallNotFound,
-	TypeNotMatch,
+	TypeNotMatch(TypeSignature, TypeSignature),
 }
 
 impl fmt::Display for Error {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(
-			f,
-			"{}",
-			match self {
-				Error::CallNotFound => "Trying to call a function that is not registered",
-				Error::TypeNotMatch => "The function is registered but the type mismatches",
-			}
-		)
+		match self {
+			Error::CallNotFound => write!(f, "Trying to call a function that is not registered"),
+			Error::TypeNotMatch(expected, found) => write!(
+				f,
+				"The function is registered but the type mismatches. Expected {}, found: {}",
+				expected, found
+			),
+		}
 	}
 }
 
 /// Register a call into the call storage.
 /// The registered call can be uniquely identified by the returned `CallId`.
-pub fn register_call<F: Fn(Args) -> R + 'static, Args: 'static, R: 'static>(f: F) -> CallId {
+pub fn register_call<F: Fn(I) -> O + 'static, I, O>(f: F) -> CallId {
+	let f = Box::new(f) as Box<dyn Fn(I) -> O>;
+	let ptr = Box::into_raw(f);
+
+	// SAFETY: transforming a wide pointer to an u128 is always safe.
+	let call = CallInfo {
+		ptr: unsafe { std::mem::transmute(ptr) },
+		type_signature: TypeSignature::new::<I, O>(),
+	};
+
 	CALLS.with(|state| {
 		let registry = &mut *state.borrow_mut();
 		let call_id = registry.len() as u64;
-		registry.insert(call_id, Box::new(CallWrapper(Box::new(f))));
+		registry.insert(call_id, call);
 		call_id
 	})
 }
 
 /// Execute a call from the call storage identified by a `call_id`.
-pub fn execute_call<Args: 'static, R: 'static>(call_id: CallId, args: Args) -> Result<R, Error> {
+pub fn execute_call<I, O>(call_id: CallId, input: I) -> Result<O, Error> {
+	let expected_type_signature = TypeSignature::new::<I, O>();
+
 	CALLS.with(|state| {
 		let registry = &*state.borrow();
 		let call = registry.get(&call_id).ok_or(Error::CallNotFound)?;
-		call.as_any()
-			.downcast_ref::<CallWrapper<Args, R>>()
-			.map(|wrapper| wrapper.0(args))
-			.ok_or(Error::TypeNotMatch)
+
+		if expected_type_signature != call.type_signature {
+			return Err(Error::TypeNotMatch(
+				expected_type_signature,
+				call.type_signature.clone(),
+			));
+		}
+
+		// SAFETY: The existance of this boxed clousure in consequent calls is ensured
```

Comment:
nit
```suggestion
		// SAFETY: The existence of this boxed closure in consequent calls is ensured
```

Reply:
This is some real inception level stuff: The existence is ensured by the `forget` call 🤯

Reply:
Realize this means that you have done a very good review. Thanks so much! 🙌🏻